programs/whirlpool/src/instructions/v2/initialize_pool.rs

```diff
@@ -0,0 +1,114 @@
+use anchor_lang::prelude::*;
+use anchor_spl::token_interface::{Mint, TokenAccount, TokenInterface};
+
+use crate::{
+  errors::ErrorCode,
+  state::*,
+  util::{is_token_badge_initialized, v2::is_supported_token_mint}
+};
+
+#[derive(Accounts)]
+#[instruction(tick_spacing: u16)]
+pub struct InitializePoolV2<'info> {
+    pub whirlpools_config: Box<Account<'info, WhirlpoolsConfig>>,
+
+    pub token_mint_a: InterfaceAccount<'info, Mint>,
+    pub token_mint_b: InterfaceAccount<'info, Mint>,
+
+    #[account(seeds = [b"token_badge", whirlpools_config.key().as_ref(), token_mint_a.key().as_ref()], bump)]
+    /// CHECK: checked in the handler
+    pub token_badge_a: UncheckedAccount<'info>,
```

Comment:
Good catch!🙏

Did you mean that the address is verified by seeds, and seeds contains whirlpools config and mint, so the remaining thing is checking if the account is initialized.

That sounds reasonable to me. Because only initialize_token_badge can make the owner of this account to WhirlpoolProgram. So we can simplify ``is_token_badge_initialized`` function. (no need to deserialize)

By using the Option, we can use Anchor to check the account, but I am concerned that it makes usage a bit more complicated since we cannot pass an uninitialized address.

Reply:
@odcheung 
I tried to remove the seeds constraint.
I found that there were no unintended malfunctions, but it was hard to notice wrong TokenBadge address  because ``is_token_badge_initialized`` would simply be false if the TokenBadge was misaddressed. I would like to leave seeds constraint as they are.

I also considered simplifying the processing of ``is_token_badge_initialized`` because the validation of seeds is sufficient, but I would like to keep ``is_token_badge_initialized`` because it is lightweight and can clearly show that the validation is sufficient.